samza-core/src/main/java/org/apache/samza/application/ApplicationDescriptorImpl.java

```diff
@@ -0,0 +1,250 @@
+   */
+  public ContextManager getContextManager() {
+    return contextManager;
+  }
+
+  /**
+   * Get the user-implemented {@link ProcessorLifecycleListenerFactory} object associated with this application
+   *
+   * @return the {@link ProcessorLifecycleListenerFactory} object
+   */
+  public ProcessorLifecycleListenerFactory getProcessorLifecycleListenerFactory() {
+    return listenerFactory;
+  }
+
+  /**
+   * Get all the {@link InputDescriptor}s to this application
+   *
+   * @return an immutable map of streamId to {@link InputDescriptor}
+   */
+  public Map<String, InputDescriptor> getInputDescriptors() {
+    return Collections.unmodifiableMap(inputDescriptors);
+  }
+
+  /**
+   * Get all the {@link OutputDescriptor}s from this application
+   *
+   * @return an immutable map of streamId to {@link OutputDescriptor}
+   */
+  public Map<String, OutputDescriptor> getOutputDescriptors() {
+    return Collections.unmodifiableMap(outputDescriptors);
+  }
+
+  /**
+   * Get all the broadcast streamIds from this application
+   *
+   * @return an immutable set of streamIds
+   */
+  public Set<String> getBroadcastStreams() {
+    return Collections.unmodifiableSet(broadcastStreams);
+  }
+
+  /**
+   * Get all the {@link TableDescriptor}s in this application
+   *
+   * @return an immutable set of {@link TableDescriptor}s
+   */
+  public Set<TableDescriptor> getTableDescriptors() {
+    return Collections.unmodifiableSet(new HashSet<>(tableDescriptors.values()));
+  }
+
+  /**
+   * Get all the unique {@link SystemDescriptor}s in this application
+   *
+   * @return an immutable set of {@link SystemDescriptor}s
+   */
+  public Set<SystemDescriptor> getSystemDescriptors() {
+    // We enforce that users must not use different system descriptor instances for the same system name
+    // when getting an input/output stream or setting the default system descriptor
+    return Collections.unmodifiableSet(new HashSet<>(systemDescriptors.values()));
+  }
+
+  /**
+   * Get the default {@link SystemDescriptor} in this application
+   *
+   * @return the default {@link SystemDescriptor}
+   */
+  public Optional<SystemDescriptor> getDefaultSystemDescriptor() {
+    return this.defaultSystemDescriptorOptional;
+  }
+
+  /**
+   * Get the {@link MetricsReporterFactory}s used in the application
+   *
+   * @return the map of {@link MetricsReporterFactory}s
+   */
+  public Map<String, MetricsReporterFactory> getMetricsReporterFactories() {
+    return Collections.unmodifiableMap(reporterFactories);
+  }
+
+  // TODO: this should go away when partitionBy() and broadcast() will also generate InputDescriptor/OutputDescriptor as well
+  // helper method to determine that there is no input/output streams added in the application yet
+  protected abstract boolean noInputOutputStreams();
+
+  // internal method to add {@link TableDescriptor} to this application
+  void addTableDescriptor(TableDescriptor tableDescriptor) {
```

Comment:
Protected actually provides a wider range of access compare to package-private. In this one, since implementation of ApplicationDescriptor is Samza framework code, we don't expect implementation class to be outside org.apache.samza.application. Hence, using package-private for those methods.